BF: revert back (remove) check for path being PathRI

yarikoptic opened the pull request:
Paths with : in them could get confused for SSHRI.  If we know that it
is a path, why to check/crash?  Original commit
86a05d0fb1a5b091676d0d430f11ae05476f18c4
which introduced that check unfortunately does not provide ANY information
on the initial problem it addresses, besides hinting possibly on it being related
to Windows

Closes #3421

mih commented:
Original reason rediscovered:

```
======================================================================
1346ERROR: datalad.distribution.tests.test_clone.test_invalid_args
1347----------------------------------------------------------------------
1348Traceback (most recent call last):
....
1387datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully [{'action': 'install', 'message': ('Failed to clone from any candidate source URL. Encountered errors per each url were: %s', (OrderedDict([('Zoidberg', "Cmd('git') failed due to: exit code(128)\n  cmdline: git clone --progress -v Zoidberg C:\\projects\\datalad\\__testhome__\\ssh:\\mars:Zoidberg [cmd.py:wait:415]")]),)), 'source_url': 'Zoidberg', 'status': 'error', 'path': 'C:\\projects\\datalad\\__testhome__\\ssh:\\mars:Zoidberg', 'type': 'dataset'}]
```

Possibly related to https://github.com/datalad/datalad/issues/2574#issuecomment-424089406 (which popped up in other issue too, but I cannot find them now)

yarikoptic commented:
yeah, deeper issue is RI resolution. looking into that ATM.

yarikoptic commented:
SSHRI issue is fixed, but I think the issue is twofold - us not checking  "higher" for a correct type of the RI... will get to that when get to work

yarikoptic commented:
ok, fix for SSHRI is in 648ada5956071b883160be76626d57e78c1e07ee
and the fix for failing test is fixing the test, see rationale in the diff and commit message of b672d4f4bfc7a4f5f15d099b3517c11e477f44c6

yarikoptic commented on datalad/support/tests/test_network.py:
```diff
@@ -272,8 +275,13 @@ def test_url_samples():
+    it is e.g. a file).
     """
-    _check_ri('example.com/path/sp1:fname', SSHRI, hostname='example.com/path/sp1', path='fname')
+    # _check_ri('example.com/path/sp1:fname', SSHRI, hostname='example.com/path/sp1', path='fname')
```
actually I will adjust here to check explicitly for being a PathRI, and bring back empty line above...

yarikoptic commented:
oh hoh -- was supposed to go against 0.11.x ... travis/appveyor would get confused... so one more amend after changing the base.

yarikoptic commented:
grr... fixed up even more, but started to wonder why I didn't generally do matching based on regex alone to decide, or at least as the 2nd check when making a decision. Oh well, may be check more on that later. Let's see how this one withstands testing

kyleam commented on datalad/support/network.py:
```diff
@@ -342,7 +342,12 @@ def _guess_ri_cls(ri):
         parts = _split_colon(ri)
-        if fields['path'] and '@' in fields['path'] or len(parts) > 1:
+        # if no illegal for hostname/username characters in the first part and
+        if not set(parts[0]).intersection(set('/\\#')) and (
```
I was surprised to the `#` given the commit message only mentioned `/` and `\`.

yarikoptic replied:
hm, indeed... will rewrite

kyleam commented on datalad/support/network.py:
```diff
@@ -342,7 +342,12 @@ def _guess_ri_cls(ri):
     if not fields['scheme'] and not fields['hostname']:
         parts = _split_colon(ri)
-        if fields['path'] and '@' in fields['path'] or len(parts) > 1:
+        # if no illegal for hostname/username characters in the first part and
```
and...? Incomplete comment?

yarikoptic replied:
thanks -- will tune up now

yarikoptic commented:
Sorry I have not adjusted description from the original one.

> Since you have it fresh in your mind, could you provide a summary?

"fresh" is already a stretch but indeed I still remember something ;)

Re "Windows failure" -- That windows failure had nothing Windows specific and was present also in Travis/locally: it was an assumption of the test that we will throw ValueError when providing a URL or URL-like where PATH is expected, instead of just returning the result record stating that clone has failed (upon git trying to clone, which it would happily do into URL or URL-like path).  The corresponding change is https://github.com/datalad/datalad/pull/3425/files#diff-41a7499eba398003a269d03345930e78L69 in commit f636b04d3bce5b7acce688ec03599cfd74a1877f which fixed that test.

**Overall summary**: 

1. stop guessing type of the RI whenever we expect a specific one, e.g. PathRI.  "URL", "URL-like" etc could all be "legit" paths and we should not second guess "if it is a legit path" by checking if it could be anything else while annotating what should already be paths.
2. fix up detection of SSHRI by relying on it having `:` (unless there is no path to follow) and no illegal for hostname characters (`/\#`) in the initial (username@hostname) part.

Makes sense?

yarikoptic commented:
force pushed after addressing comments, extending regex for sshri also to include `#` (to make it more consistent with the detection code), squashing two related commits

kyleam commented:
> Re "Windows failure" -- That windows failure had nothing Windows specific and was present also in Travis/locally: 

The revert of the PathRI check for resolve_path() (b94fc33f0) did _not_ lead to failures on Travis [0], even though it led to a failure on Windows [1]. That failing test, test_clone.py:test_invalid_args, also passed for me locally when I run it on b94fc33f0.


[0]: https://travis-ci.org/datalad/datalad/builds/533129300
[1]: https://github.com/datalad/datalad/pull/3425#issuecomment-492923959

> **Overall summary**: 
>
> 1. stop guessing type of the RI whenever we expect a specific one, e.g. PathRI.  "URL", "URL-like" etc could all be "legit" paths and we should not second guess "if it is a legit path" by checking if it could be anything else while annotating what should already be paths.
> 2. fix up detection of SSHRI by relying on it having `:` (unless there is no path to follow) and no illegal for hostname characters (`/\#`) in the initial (username@hostname) part.

Thanks, that matches what I got from the individual commits.  But I'm still missing how the new changes relate to the removal of the PathRI check from resolve_path().  Apparently that Windows failure may be the reason it was added, but I don't quite get how this series fixes that (although, by the end, appveyor is happy).

Anyway, no need to hold this up if I'm just being too obtuse or lazy to connect the obvious pieces.

kyleam commented:
> force pushed after addressing comments, extending regex for sshri also to include `#` (to make it more consistent with the detection code), squashing two related commits

Thanks!

<details>
<summary>range-diff</summary>

```
1:  98a057e41 ! 1:  2e5e9b1f7 BF: Forbid having / or \ in the hostname of the SSHRI while detecting
    @@ -1,6 +1,6 @@
     Author: Yaroslav Halchenko <debian@onerussian.com>
     
    -    BF: Forbid having / or \ in the hostname of the SSHRI while detecting
    +    BF: Check for /, \ or # in the username@hostname part while detecting SSHRI
     
         ref: see https://github.com/datalad/datalad/issues/3421
     
    @@ -12,8 +12,13 @@
          if not fields['scheme'] and not fields['hostname']:
              parts = _split_colon(ri)
     -        if fields['path'] and '@' in fields['path'] or len(parts) > 1:
    -+        if fields['path'] and '@' in fields['path'] \
    -+                or (len(parts) > 1 and not set(parts[0]).intersection(set('/\\#'))):
    ++        # if no illegal for username@hostname characters in the first part and
    ++        # we either had username@hostname or multiple :-separated parts
    ++        if not set(parts[0]).intersection(set('/\\#')) and (
    ++                fields['path'] and
    ++                '@' in fields['path'] or
    ++                len(parts) > 1
    ++        ):
                  # user@host:path/sp1
                  # or host_name: (hence parts check)
                  # TODO: we need a regex to catch those really, parts check is not suff
    @@ -22,7 +27,7 @@
          )
      
     -    _REGEX = re.compile(r'((?P<username>\S*)@)?(?P<hostname>[^:]+)(\:(?P<path>.*))?$')
    -+    _REGEX = re.compile(r'((?P<username>\S*)@)?(?P<hostname>[^/\\:]+)(\:(?P<path>.*))?$')
    ++    _REGEX = re.compile(r'((?P<username>\S*)@)?(?P<hostname>[^#/\\:]+)(\:(?P<path>.*))?$')
      
          @classmethod
          def _normalize_fields(cls, fields):
    @@ -52,6 +57,7 @@
          """
     -    _check_ri('example.com/path/sp1:fname', SSHRI, hostname='example.com/path/sp1', path='fname')
     +    _check_ri('e.com/p/sp:f', PathRI, localpath='e.com/p/sp:f', path='e.com/p/sp:f')
    ++    _check_ri('user@someho.st/mydir', PathRI, localpath='user@someho.st/mydir', path='user@someho.st/mydir')
      
          # SSHRIs have .port, but it is empty
          eq_(SSHRI(hostname='example.com').port, '')
2:  f636b04d3 = 2:  60bfefe87 BF(TST): do not assume user naiveness - treat any url-like looking path as a path
3:  6831aa7de < -:  --------- BF: generally check for no inappropriate characters in the ssh RI beginning
```

</details>

yarikoptic commented:
hm, the only failing run is for -neuroimaging:
```shell
$ http_proxy= PATH=$PWD/../tools/coverage-bin:$PATH $NOSE_WRAPPER `which nosetests` $NOSE_OPTS -A "$NOSE_SELECTION_OP($NOSE_SELECTION)" --with-doctest --with-cov --cover-package datalad --logging-level=INFO $TESTS_TO_PERFORM
.........SS.E.S.
======================================================================
ERROR: datalad_neuroimaging.tests.test_procedure.test_bids_procedure
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 615, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad_neuroimaging/tests/test_procedure.py", line 13, in test_bids_procedure
    ds.run_procedure(['cfg_bids'])
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/distribution/dataset.py", line 526, in apply_func
    return f(**kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 492, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 480, in return_func
    results = list(results)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 468, in generator_func
    msg="Command did not complete successfully")
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully [{'action': 'run_procedure', 'status': 'impossible', 'refds': '/tmp/datalad_temp_test_bids_procedure0ucvyzgv', 'path': 'cfg_bids', 'message': "Cannot find procedure with name 'cfg_bids'"}]
```

yarikoptic commented:
I have tried to reproduce locally by running tests of datalad-neuroimaging -- do not see anything suspicious.   and happens also for other PRs such as https://github.com/datalad/datalad/pull/3430 .

So unrelated to this one.  Unless more objections are heard, will merge tomorrow morning

kyleam commented:
Not a reason to not merge, since the end result seems fine test-wise, but it'd be nice if you'd acknowledge the first part of my comment here: https://github.com/datalad/datalad/pull/3425#issuecomment-493508896

Based on your claim that the test_clone.py:test_invalid_args failure at b94fc33f0 was not appveyor-specific, I'd say you don't fully understand some aspect of these changes.

yarikoptic commented:
Oh, indeed. I i just saw that Travis also had a red cross, and assumed that the failure was the same... But I will check how that clone test managed to pass in that state when I get to the laptop.

yarikoptic commented:
ha!   the test was not failing on Linux because ValueError **was** raised just at a different point but similarly (RI resolution to SSHRI) inflicted in a different spot:

```
$> python -m nose --pdb -s -v datalad/distribution/tests/test_clone.py:test_invalid_args
datalad.distribution.tests.test_clone.test_invalid_args ... > /home/yoh/proj/datalad/datalad/datalad/support/network.py(476)localpath()
-> raise ValueError("%s points to remote location" % self)
(Pdb) p self
SSHRI(hostname='/home/yoh/proj/datalad/datalad/ssh', path='/mars:Zoidberg')
(Pdb) bt
  /usr/lib/python2.7/unittest/case.py(329)run()
-> testMethod()
  /usr/lib/python2.7/dist-packages/nose/case.py(197)runTest()
-> self.test(*self.arg)
  /home/yoh/proj/datalad/datalad/datalad/tests/utils.py(615)newfunc()
-> return t(*(arg + (filename,)), **kw)
  /home/yoh/proj/datalad/datalad/datalad/tests/utils.py(615)newfunc()
-> return t(*(arg + (filename,)), **kw)
  /home/yoh/proj/datalad/datalad/datalad/tests/utils.py(615)newfunc()
-> return t(*(arg + (filename,)), **kw)
  /home/yoh/proj/datalad/datalad/datalad/distribution/tests/test_clone.py(70)test_invalid_args()
-> clone('Zoidberg', path='ssh://mars:Zoidberg')
  /home/yoh/proj/datalad/datalad/datalad/interface/utils.py(492)eval_func()
-> return return_func(generator_func)(*args, **kwargs)
  /home/yoh/proj/datalad/datalad/datalad/interface/utils.py(480)return_func()
-> results = list(results)
  /home/yoh/proj/datalad/datalad/datalad/interface/utils.py(429)generator_func()
-> result_renderer, result_xfm, _result_filter, **_kwargs):
  /home/yoh/proj/datalad/datalad/datalad/interface/utils.py(522)_process_results()
-> for res in results:
  /home/yoh/proj/datalad/datalad/datalad/distribution/clone.py(182)__call__()
-> destination_dataset = Dataset(path)
  /home/yoh/proj/datalad/datalad/datalad/support/repo.py(140)__call__()
-> id_, new_args, new_kwargs = cls._flyweight_id_from_args(*args, **kwargs)
  /home/yoh/proj/datalad/datalad/datalad/distribution/dataset.py(140)_flyweight_id_from_args()
-> path_ = RI(path_).localpath
  /home/yoh/proj/datalad/datalad/datalad/support/network.py(541)__getattribute__()
-> return super(RI, self).__getattribute__(item)
> /home/yoh/proj/datalad/datalad/datalad/support/network.py(476)localpath()
-> raise ValueError("%s points to remote location" % self)
```
So fixing the SSHRI parsing made it no longer fail here since it was no longer identified as an SSHRI.

kyleam commented:
> ha!   the test was not failing on Linux because ValueError **was** raised just at a different point but similarly (RI resolution to SSHRI) inflicted in a different spot:

I see.  Thanks for digging.
